sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -84,26 +91,98 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
   return Program;
 }
 
+template<typename RetT>
+RetT *waitUntilBuilt(const std::shared_ptr<context_impl> &Ctx,
+                     context_impl::WithBuildState<RetT> *WithBuildState) {
+  // condition_variable.wait() call is wrapped with a while due to common
+  // cond_var for building all kernels and programs
+  int state = WithBuildState->State.load();
+  while (state != BS_Built && state != BS_BuildFailed) {
+    Ctx->waitProgramBuild([WithBuildState]() {
+      auto state = WithBuildState->State.load();
+
+      return state == BS_Built || state == BS_BuildFailed;
+    });
+
+    state = WithBuildState->State.load();
+  }
+
+  // expected to be null if state eq BuildFailed
+  return WithBuildState->Ptr.load();
+}
+
 RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
                                                     const context &Context) {
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
-  std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
-      Ctx->getCachedPrograms();
-  auto It = CachedPrograms.find(M);
-  if (It != CachedPrograms.end())
-    return It->second;
 
+  using PiProgramT = context_impl::PiProgramT;
+  using ProgramWithBuildState = context_impl::WithBuildState<PiProgramT>;
+
+  ProgramWithBuildState *ProgramWithState;
+  bool InsertionTookPlace;
+
+  // put atomic pointer in program cache as data with nullptr before
+  // the program is actually built
+  {
+    auto Locked = Ctx->acquireCachedPrograms();
+    auto &CachedPrograms = Locked.get();
+
+    auto Inserted = CachedPrograms.emplace(std::piecewise_construct,
+                                           std::forward_as_tuple(M),
+                                           std::forward_as_tuple(nullptr,
+                                                                 BS_None));
+
+    InsertionTookPlace = Inserted.second;
+    ProgramWithState = &Inserted.first->second;
+  }
+
+  // any thread which will find nullptr in cache will wait until the pointer
+  // is not null anymore
+  if (!InsertionTookPlace) {
+    return waitUntilBuilt(Ctx, ProgramWithState);
+  }
+
+  {
+    int Expected = BS_None;
```

Comment:
```suggestion
    const int Expected = BS_None;
```

Reply:
atomic requires a ref to non-const var for `expected` in all of its CAS methods.